File: server/routes/recipes.ts
import { RequestHandler } from "express";
import { prisma } from "../lib/prisma";
import { optionalAuth } from "../middleware/auth";

export const listRecipes: RequestHandler = async (req, res) => {
  try {
    const {
      cuisine,
      tags,
      minRating,
      search,
      limit = "50",
      offset = "0",
    } = req.query;

    const where: any = {};

    if (cuisine) {
      where.cuisine = cuisine as string;
    }

    if (tags) {
      const tagArray = Array.isArray(tags) ? tags : [tags];
      where.tags = {
        hasSome: tagArray as string[],
      };
    }

    if (minRating) {
      where.avgRating = {
        gte: parseFloat(minRating as string),
      };
    }

    if (search) {
      where.OR = [
        { name: { contains: search as string, mode: "insensitive" } },
        { description: { contains: search as string, mode: "insensitive" } },
      ];
    }

    const recipes = await prisma.recipe.findMany({
      where,
      take: parseInt(limit as string),
      skip: parseInt(offset as string),
      orderBy: { avgRating: "desc" },
      include: {
        ingredients: {
          include: {
            ingredient: true,
          },
        },
      },
    });

    res.json(recipes);
  } catch (error) {
    res.status(500).json({ error: "Failed to fetch recipes" });
  }
};

export const getRecipe: RequestHandler = async (req, res) => {
  try {
    const { id } = req.params;

    const recipe = await prisma.recipe.findUnique({
      where: { id },
      include: {
        ingredients: {
          include: {
            ingredient: true,
          },
        },
        feedbacks: {
          take: 10,
          orderBy: { createdAt: "desc" },
          include: {
            user: {
              select: {
                id: true,
                name: true,
                avatarUrl: true,
              },
            },
          },
        },
      },
    });

    if (!recipe) {
      return res.status(404).json({ error: "Recipe not found" });
    }

    res.json(recipe);
  } catch (error) {
    res.status(500).json({ error: "Failed to fetch recipe" });
  }
};

export const listRecipeHandler = [optionalAuth, listRecipes];
export const getRecipeHandler = [optionalAuth, getRecipe];

File: server/routes/recipes.py
"""
Recipe routes
"""
from fastapi import APIRouter, HTTPException, Query, Depends
from typing import Optional, List
from server.services.supabase_client import get_supabase_client
from server.middleware.auth import optional_auth

router = APIRouter()

@router.get("")
async def list_recipes(
    cuisine: Optional[str] = Query(None),
    tags: Optional[str] = Query(None),
    min_rating: Optional[float] = Query(None),
    search: Optional[str] = Query(None),
    limit: int = Query(50, ge=1, le=100),
    offset: int = Query(0, ge=0),
    user: Optional[dict] = Depends(optional_auth)
):
    """List recipes with filters"""
    supabase = get_supabase_client()
    
    try:
        query = supabase.table("recipes").select("*")
        
        if cuisine:
            query = query.eq("cuisine", cuisine)
        
        if min_rating:
            query = query.gte("avg_rating", min_rating)
        
        if search:
            query = query.or_(f"name.ilike.%{search}%,description.ilike.%{search}%")
        
        if tags:
            tag_list = tags.split(",")
            for tag in tag_list:
                query = query.contains("tags", [tag])
        
        response = query.order("avg_rating", desc=True).range(offset, offset + limit - 1).execute()
        
        return response.data or []
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/{recipe_id}")
async def get_recipe(
    recipe_id: str,
    user: Optional[dict] = Depends(optional_auth)
):
    """Get recipe details"""
    supabase = get_supabase_client()
    
    try:
        response = supabase.table("recipes").select("*").eq("id", recipe_id).execute()
        
        if not response.data:
            raise HTTPException(status_code=404, detail="Recipe not found")
        
        # Get feedback if available
        feedback_response = supabase.table("feedback").select("*").eq("recipe_id", recipe_id).order("created_at", desc=True).limit(10).execute()
        
        recipe = response.data[0]
        recipe["feedbacks"] = feedback_response.data or []
        
        return recipe
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

